### PR TITLE
Searchable snapshots mounted in the 'hot' phase should be pinned to hot nodes

### LINKED ISSUE #68635: ILM optimized migrate for searchable snapshot action
Today, the migrate action will execute before the searchable snapshot action.
If tiers are separated on different sets of nodes, this means that the shards are
first transferred to the new tier (cold for instance) and then afterwards mounted
as searchable snapshots on that same tier.

It seems we could be more optimal here by not doing the migrate step on the
existing index and instead assign the mounted index directly to the target tier.

### LINKED ISSUE #72340: Searchable snapshots mounted in the 'hot' phase should be pinned to hot nodes
We added the ability to allow mounting an index from an ILM policy in the `hot` phase after rollover, however, these indices are automatically sent to the cold nodes. We need to change ILM such that we pass in `_tier_preference: data_hot` when they are mounted in the hot phase.

We'll also want to ensure they are migrated between phases appropriately.

### PR DESCRIPTION
**Pull request by joegallo**
Closes #72340

Some bits of this have the effect of reverting or altering some parts of #68861 (which itself closed #68635) -- so I added @andreidan as a reviewer to make sure he agrees with where we landed here.

### CODE REVIEW

**Comment by joegallo** [comment]
`testSearchableSnapshotActionOverridesMigrateAction` is going to fail on this CI run (and indeed it also [fails locally](https://gradle-enterprise.elastic.co/s/vyo4jcysfk7rq)) -- I need to figure out what to do about that one.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java [hunk 1]
```diff
@@ -130,11 +129,8 @@ public boolean isSafeAction() {
         }
     }
 
-    static boolean skipMigrateAction(String phase, IndexMetadata indexMetadata) {
-        // if the index is a searchable snapshot we skip the migrate action (as mounting an index as searchable snapshot
-        // configures the tier allocation preference), unless we're in the frozen phase
-        return (indexMetadata.getSettings().get(LifecycleSettings.SNAPSHOT_INDEX_NAME) != null)
-            && (phase.equals(FROZEN_PHASE) == false);
+    private static boolean skipMigrateAction(IndexMetadata indexMetadata) {
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java:132]: Can you add a comment about why we only skip for partially-mounted indices here?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MountSnapshotStep.java [hunk 2]
```diff
@@ -101,10 +103,14 @@ void performDuringNoSnapshot(IndexMetadata indexMetadata, ClusterState currentCl
             indexName = snapshotIndexName;
         }
 
+        Settings.Builder settingsBuilder = Settings.builder();
+        settingsBuilder.put(IndexSettings.INDEX_CHECK_ON_STARTUP.getKey(), Boolean.FALSE.toString());
+        if (TimeseriesLifecycleType.HOT_PHASE.equals(this.getKey().getPhase())) {
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MountSnapshotStep.java:108]: Can you add a comment about why this is necessary here?

**Review by dakrone — APPROVED** [review]
LGTM, I left two really minor requests for comments so that we don't lose our reasoning down the line next time we look at this code.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java (hunk 1 above)
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java:132]: I believe this method is quite shallow so we should drop it altogether? I believe its usage should be replaced by 
```
SearchableSnapshotsConstants.isPartialSearchableSnapshotIndex(indexMetadata.getSettings());
```
with the appropriate comment about why we only skip in this particular case

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java [hunk 3]
```diff
@@ -109,7 +108,7 @@ public boolean isSafeAction() {
 
             BranchingStep conditionalSkipActionStep = new BranchingStep(preMigrateBranchingKey, migrationKey, nextStepKey,
                 (index, clusterState) -> {
-                    if (skipMigrateAction(phase, clusterState.metadata().index(index))) {
+                    if (skipMigrateAction(clusterState.metadata().index(index))) {
                         String policyName =
                             LifecycleSettings.LIFECYCLE_NAME_SETTING.get(clusterState.metadata().index(index).getSettings());
                         logger.debug("[{}] action is configured for index [{}] in policy [{}] which is already mounted as a searchable " +
```
Comment (andreidan) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java:114]: Shall we update the logging message to refer to a "partially mounted index"?

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/SearchableSnapshotActionIT.java [hunk 4]
```diff
@@ -503,6 +538,7 @@ public void testSearchableSnapshotActionOverridesMigrateAction() throws Exceptio
         indexDocument(client(), dataStream, true);
         String firstGenIndex = DataStream.getDefaultBackingIndexName(dataStream, 1L);
         Map<String, Object> indexSettings = getIndexSettingsAsMap(firstGenIndex);
+        // the searchable snapshot is allocated to the hot tier
```
Comment (andreidan) [inline x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/SearchableSnapshotActionIT.java:541]: I think this is incorrect? This is asserting the managed index which hasn't yet been mounted as a searchable snapshot (we haven't rolled over yet). The assertion passes due to our default allocation of new indices to the hot tier. 

I'm not sure we need this IT anymore as the only time a mounted index will override the migrate action is in the frozen phase (the only place where we allow a partially mounted index and when we skip injecting the migrate action) - the allocation routing configured by the mounted index and the migrate action are identical so we can't distinguish who set it anymore. 

I suggest we convert this IT to a unit test for `MigrateAction#toSteps` in `MigrateActionTests` to make sure the migrate action is injected only when appropriate.

What do you think?

**Review by andreidan — COMMENTED** [review]
Thanks for working on this @joegallo 

I've left a couple of comments

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java (hunk 1 above)
Comment (joegallo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java]: Inlined in 6bd08fec59b301ff63794cbcb443e008016b119b, commented in 8695b510e32c1972d76492052adf5e680970f10d.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java (hunk 3 above)
Comment (joegallo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java]: Done, 0b9f44a577b384a2f5c1c8a07a6a716c5bcc9a6a

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MountSnapshotStep.java (hunk 2 above)
Comment (joegallo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MountSnapshotStep.java:109]: Sure thing, 21cabd43d71dd6e77c4d9c1729f05fbe3c7c75f9

**Review by dakrone — APPROVED** [review]
LGTM after addressing Andrei's comment

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/SearchableSnapshotActionIT.java (hunk 4 above)
Comment (joegallo) [reply x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/SearchableSnapshotActionIT.java]: Makes sense to me, @andreidan. What do you think of 46e4ba4?

**Review by andreidan — APPROVED** [review]
LGTM, thanks for working on this @joegallo

**Pull request merged**